gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -294,49 +328,145 @@ public void run() {
                                 UInt32Value.newBuilder().setValue(objectHasher.hash().asInt())));
                   }
                 }
-                requestObserver.onNext(requestBuilder.build());
-
-                if (objectFinalized) {
-                  // Close the request half of the streaming RPC.
-                  requestObserver.onCompleted();
+                if (dataChunkMap.size() == numberOfRequestsToRetain) {
+                  dataChunkMap.remove(dataChunkMap.firstKey());
+                  retryMap.remove(dataChunkMap.firstKey());
                 }
+                dataChunkMap.put(writeOffset, chunkData);
+                return requestBuilder;
               }
 
               private ByteString readRequestData() throws IOException {
-                // Mark the input stream in case this request fails so that read can be recovered
-                // from where it's marked.
-                pipeSource.mark(MAX_BYTES_PER_MESSAGE);
                 ByteString data =
                     ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
-
-                objectFinalized =
-                    data.size() < MAX_BYTES_PER_MESSAGE || pipeSource.available() <= 0;
+                pipeSource.mark(1);
+                objectFinalized = data.size() < MAX_BYTES_PER_MESSAGE || pipeSource.read() == -1;
+                pipeSource.reset();
                 return data;
               }
+
+              // Handles the case when a writeOffset of data read previously is being processed.
+              // This happens if a transient failure happens while uploading, and can be resumed by
+              // querying the current committed offset.
+              private InsertObjectRequest.Builder buildRequestFromCachedDataChunk(
+                  TreeMap<Long, ByteString> dataChunkMap) {
+                // Resume will only work if the first request builder in the cache carries an offset
+                // not greater than the current writeOffset.
+                InsertObjectRequest.Builder requestBuilder = null;
+                if (dataChunkMap.size() > 0 && dataChunkMap.firstKey() <= writeOffset) {
+                  long prevWriteOffset = 0;
+                  ByteString prevDataChunk = null;
+                  for (Map.Entry<Long, ByteString> entry : dataChunkMap.entrySet()) {
+                    // Take the previous entry if it's greater than current writeOffset.
+                    if (entry.getKey() == writeOffset) {
```

Comment:
Why do we need to handle the case where entry.getKey() == writeOffset separately from the case where entry.getKey() > writeOffset? I think you just need to iterate until you get to the first entry whose offset is <= last persisted offset + write size (which means the cached request contains the offset where we want to resume writing). I think the code would be more readable if you have this one condition instead of two.

Reply:
I am not sure I understand correctly "whose offset is <= last persisted offset + write size". I think regardless we need to diff btw '<' and '=='. So just to be clear, each entry in the map means <write offset, datachunk that will be written from this write offset>. So ideally the new writeOffset that comes with a new insert would be larger than all keys in the caching map. If not, there are two cases 1) new writeOffset is even smaller than the first key of the caching map, in which case we will just consider as non-resumable and as a in-transient error 2) this is the case we can resume from the caching map, if the new writeOffset is equal to any of previous key that means we can redo insertion with exactly that offset and the corresponding data chunk. However if there is no exact match, we want to find the first key(offset) in the map that is greater that this new offset but we do not use this one to redo insert because there will be piece of last data chunk that's gonna be missing(remember the current new writeOffset means committed size we queried from server). Instead we want to use the entry <offset, datachunk> one before the offset that's larger than new writeOffset. Hope this helps to explain...